File: frontend/src/lib/racebox/ubxParser.ts
/**
 * RaceBox Mini UBX-like packet parser.
 * Protocol: 0xB5 0x62 | class | id | lenL lenH | payload | CK_A CK_B
 * RaceBox Data Message: class=0xFF, id=0x01, payload=80 bytes.
 */

export interface RaceBoxSample {
  timestamp: number;       // performance.now() at parse time
  lat: number;             // degrees
  lon: number;             // degrees
  altitudeM: number;       // meters MSL
  speedMms: number;        // mm/s ground speed
  speedKmh: number;        // km/h
  headingDeg: number;      // degrees
  gForceX: number;         // G (lateral)
  gForceY: number;         // G (longitudinal)
  gForceZ: number;         // G (vertical)
  fixType: number;         // 0=none, 2=2D, 3=3D
  numSatellites: number;
  batteryPercent: number;
}

const SYNC1 = 0xb5;
const SYNC2 = 0x62;
const RACEBOX_CLASS = 0xff;
const RACEBOX_DATA_ID = 0x01;
const HEADER_LEN = 6; // sync(2) + class(1) + id(1) + len(2)
const CHECKSUM_LEN = 2;

export class UbxParser {
  private buffer = new Uint8Array(1024);
  private pos = 0;

  /** Feed raw BLE notification bytes. Returns parsed samples (0 or more). */
  feed(data: DataView): RaceBoxSample[] {
    // Append to buffer
    const incoming = new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
    if (this.pos + incoming.length > this.buffer.length) {
      // Grow buffer
      const newBuf = new Uint8Array(this.buffer.length * 2);
      newBuf.set(this.buffer.subarray(0, this.pos));
      this.buffer = newBuf;
    }
    this.buffer.set(incoming, this.pos);
    this.pos += incoming.length;

    const samples: RaceBoxSample[] = [];

    // Try to extract complete packets
    while (this.pos >= HEADER_LEN) {
      // Find sync header
      const syncIdx = this.findSync();
      if (syncIdx < 0) {
        this.pos = 0;
        break;
      }
      if (syncIdx > 0) {
        // Discard bytes before sync
        this.buffer.copyWithin(0, syncIdx, this.pos);
        this.pos -= syncIdx;
      }

      if (this.pos < HEADER_LEN) break;

      const payloadLen = this.buffer[4] | (this.buffer[5] << 8);
      const packetLen = HEADER_LEN + payloadLen + CHECKSUM_LEN;

      if (this.pos < packetLen) break; // Need more data

      // Verify checksum (Fletcher-8 over class+id+len+payload)
      let ckA = 0, ckB = 0;
      for (let i = 2; i < HEADER_LEN + payloadLen; i++) {
        ckA = (ckA + this.buffer[i]) & 0xff;
        ckB = (ckB + ckA) & 0xff;
      }

      const msgClass = this.buffer[2];
      const msgId = this.buffer[3];

      if (ckA === this.buffer[HEADER_LEN + payloadLen] &&
          ckB === this.buffer[HEADER_LEN + payloadLen + 1] &&
          msgClass === RACEBOX_CLASS && msgId === RACEBOX_DATA_ID &&
          payloadLen === 80) {
        const sample = this.parsePayload(this.buffer, HEADER_LEN);
        if (sample) samples.push(sample);
      }

      // Consume packet
      this.buffer.copyWithin(0, packetLen, this.pos);
      this.pos -= packetLen;
    }

    return samples;
  }

  private findSync(): number {
    for (let i = 0; i <= this.pos - 2; i++) {
      if (this.buffer[i] === SYNC1 && this.buffer[i + 1] === SYNC2) return i;
    }
    return -1;
  }

  private parsePayload(buf: Uint8Array, offset: number): RaceBoxSample | null {
    const dv = new DataView(buf.buffer, buf.byteOffset + offset, 80);

    const fixType = dv.getUint8(20);
    const numSat = dv.getUint8(23);
    const lon = dv.getInt32(24, true) / 1e7;
    const lat = dv.getInt32(28, true) / 1e7;
    const altMsl = dv.getInt32(36, true) / 1000; // mm -> m
    const speedMms = dv.getInt32(48, true);       // mm/s (signed)
    const heading = dv.getInt32(52, true) / 1e5;  // deg
    const battery = dv.getUint8(67) & 0x7f;       // lower 7 bits
    const gfx = dv.getInt16(68, true) / 1000;     // milli-g -> G
    const gfy = dv.getInt16(70, true) / 1000;
    const gfz = dv.getInt16(72, true) / 1000;

    return {
      timestamp: performance.now(),
      lat, lon,
      altitudeM: altMsl,
      speedMms: Math.abs(speedMms),
      speedKmh: Math.abs(speedMms) * 3.6 / 1000,
      headingDeg: heading,
      gForceX: gfx,
      gForceY: gfy,
      gForceZ: gfz,
      fixType,
      numSatellites: numSat,
      batteryPercent: battery,
    };
  }

  reset() {
    this.pos = 0;
  }
}

File: BoxBoxNow/BoxBoxNow/Services/UbxParser.swift
import Foundation
import QuartzCore

/// Parser for the RaceBox Mini UBX-like BLE data stream.
///
/// Protocol (matches `frontend/src/lib/racebox/ubxParser.ts`):
///   `0xB5 0x62 | class | id | lenL lenH | payload | CK_A CK_B`
/// RaceBox Data Message: class=0xFF, id=0x01, payload=80 bytes.
///
/// NOTE: this is the RaceBox proprietary message, NOT the standard
/// UBX-NAV-PVT (class=0x01, id=0x07, payload=92 bytes). The real device
/// only sends the 0xFF/0x01 message, so matching on anything else means
/// we never emit samples — which previously left the calibrator stuck
/// at 0% because gForce data never reached it.
final class UbxParser {
    var onParsed: ((GPSSample) -> Void)?

    private var buffer = Data()
    private let syncA: UInt8 = 0xB5
    private let syncB: UInt8 = 0x62
    private let raceboxClass: UInt8 = 0xFF
    private let raceboxDataId: UInt8 = 0x01
    private let expectedPayloadLen: Int = 80

    func feed(_ data: Data) { buffer.append(data); parse() }
    func reset() { buffer.removeAll() }

    private func parse() {
        while buffer.count >= 8 {
            guard let syncIdx = findSync() else { buffer.removeAll(); return }
            if syncIdx > 0 { buffer = Data(buffer.dropFirst(syncIdx)) }
            guard buffer.count >= 6 else { return }

            let payloadLen = Int(buffer[4]) | (Int(buffer[5]) << 8)
            let total = 6 + payloadLen + 2
            guard buffer.count >= total else { return }

            let packet = Data(buffer.prefix(total))
            if verifyChecksum(packet)
                && buffer[2] == raceboxClass
                && buffer[3] == raceboxDataId
                && payloadLen == expectedPayloadLen {
                if let sample = parseRaceBoxData(Data(packet[6..<(6 + payloadLen)])) {
                    onParsed?(sample)
                }
            }
            buffer = Data(buffer.dropFirst(total))
        }
    }

    private func findSync() -> Int? {
        guard buffer.count >= 2 else { return nil }
        for i in 0..<(buffer.count - 1) {
            if buffer[i] == syncA && buffer[i + 1] == syncB { return i }
        }
        return nil
    }

    private func verifyChecksum(_ p: Data) -> Bool {
        var a: UInt8 = 0, b: UInt8 = 0
        for i in 2..<(p.count - 2) { a = a &+ p[i]; b = b &+ a }
        return a == p[p.count - 2] && b == p[p.count - 1]
    }

    /// Parse the 80-byte RaceBox payload. Offsets mirror the web parser.
    private func parseRaceBoxData(_ p: Data) -> GPSSample? {
        guard p.count >= 80 else { return nil }

        // Read little-endian integers from the copied payload. We use a
        // local copy via withUnsafeBytes to avoid alignment surprises on
        // ARM when reading int32/int16 fields from arbitrary offsets.
        func i4(_ o: Int) -> Int32 {
            var v: Int32 = 0
            _ = withUnsafeMutableBytes(of: &v) { dst in
                p.withUnsafeBytes { src in
                    memcpy(dst.baseAddress!, src.baseAddress!.advanced(by: o), 4)
                }
            }
            return Int32(littleEndian: v)
        }
        func i2(_ o: Int) -> Int16 {
            var v: Int16 = 0
            _ = withUnsafeMutableBytes(of: &v) { dst in
                p.withUnsafeBytes { src in
                    memcpy(dst.baseAddress!, src.baseAddress!.advanced(by: o), 2)
                }
            }
            return Int16(littleEndian: v)
        }
        func u1(_ o: Int) -> UInt8 { p[o] }

        let fixType = Int(u1(20))
        let numSat = Int(u1(23))
        let lon = Double(i4(24)) * 1e-7
        let lat = Double(i4(28)) * 1e-7
        let alt = Double(i4(36)) / 1000.0          // mm → m
        let spdMms = Double(i4(48))                // mm/s (signed)
        let hdg = Double(i4(52)) * 1e-5            // deg
        let battery = Int(u1(67) & 0x7f)           // lower 7 bits
        let gfx = Double(i2(68)) / 1000.0          // milli-G → G
        let gfy = Double(i2(70)) / 1000.0
        let gfz = Double(i2(72)) / 1000.0

        let spdKmh = abs(spdMms) * 3.6 / 1000.0

        return GPSSample(
            timestamp: CACurrentMediaTime(),
            lat: lat, lon: lon, altitudeM: alt,
            speedKmh: spdKmh, headingDeg: hdg,
            gForceX: gfx, gForceY: gfy, gForceZ: gfz,
            fixType: fixType, numSatellites: numSat, batteryPercent: battery
        )
    }
}
